#August 2015 
#Author: FBB

#reproduce the result in http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 


In [32]:
import os
import sys
import numpy as np
import pylab as pl


import json
import os
s = json.load( open(os.getenv("PUI2015")+"/fbb_matplotlibrc.json") )
pl.rcParams.update(s)

%pylab inline

from IPython.display import Image
Image(filename='../plotsforclasses/NYCReentryprogram_title.png')



TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'

In [ ]:
Image(filename='../plotsforclasses/NYCReentryprogram.png')

#NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

#$H_0: P_0 - P_1 \geq$    0
    
#$H_a: P_0 - P_1 $< 0    
    
    
#$\alpha$ = 0.05    

##this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (Bernoulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):
#$P_0=0.035, P_1=0.701$

In [ ]:

alpha=0.05
#we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0=3.5*0.01 
P_1=70.1*0.01

n_0=409
n_1=564

#lets get the counts by multiplying by the sample size
Nt_0=P_0*n_0
Nt_1=P_1*n_1


#2 samples, categorical data
#TWO OPTIONS z test, or chi square test.  

#START WITH Z TEST

##the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality
must define the sample standard deviation 

In [ ]:
#define the sample proportion first
sp=(P_0*n_0+P_1*n_1)/(n_1+n_0)
print sp

#standard deviation of the sampling distribution the distribution is Binomial, the binomail stdev is 

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!): 

$\sqrt{\frac{p(1 - p)}{n}}$
    

In [ ]:
# i am goonna create a little one line function to calculate the standard dev, it is not really needed, but just to show you how you do such a thing
sp_stdev= lambda p, n: np.sqrt(p*(1-p)/n)

sp_stdev_2y=sp_stdev(P_0,n_0+n_1)
print P_0, n_0, n_1, sp_stdev_2y

#z score: how many standard deviation away from the population parameter is my statistic?

#$z=\frac{P_1-P_0}{\sigma}$

In [ ]:
zscore = lambda p0, p1, s : (p0-p1)/s
z_2y = zscore(P_1, P_0, sp_stdev_2y)
print z_2y


In [ ]:
Image('http://intersci.ss.uci.edu/wiki/images/3/3a/Normal01.jpg')

##if $p<\alpha$ : reject H0

##IMPORTANT!! note that this P is not the p-value, but 
##p-value=1-P

In [ ]:
##p-value for employment after 2 years
p_2y=1-0.9998

def report_result(p,a):
    print 'is the p value {0:.2f} smaller than the critical value {1:.2f}? '.format(p,a)
    if p<a:
        print "YES!"
    else: print "NO!"
    
    print 'the Null hypothesis is {}'.format( 'rejected' if p<a  else 'not rejected') 

    
report_result(p_2y,alpha)

#!!!!!!!!TODO FOR YOU


##what if we used the values for where the former inmate was or was not ever employed in a CEO transitional job 3 years after release (5th row of the table)? 

#$P_0 = 0.282, P_1= 0.272$

#look up data table and insert the appropriate values to get the appropriate result!

In [ ]:
P_0=28.2*0.01 
P_1=27.2*0.01 
Nt_0=P_0*n_0
Nt_1=P_1*n_1
sp_stdev_3y=sp_stdev(P_0,n_0+n_1)
print P_0, P_1 
print n_0, n_1 
print sp_stdev_3y

In [ ]:
z_3y = zscore(P_1, P_0, sp_stdev_3y)
print z_3y

In [ ]:
p_3y=1-0.7549
    
report_result(p_3y,alpha)

#Now lets do it with the $\chi^2$ test

#this can also be done with the $\chi^2$ distribution, see "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

##the chisq statistics tests a number against the distribution of the following quantity:

$$\chi^2 = \Sigma \frac{(observation - expectation)^2}{expectation}$$


if we talk about sample fractions  that is 

$$\chi^2 = \Sigma \frac{(f_{observed} - f_{expectated})^2}{f_{expected}}$$

##turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

##the trickiest part (not that tricky) is to figure out the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example:

|employed in subsadized job |     employed   | not employes   |                   
|---------------------------|----------------|------------------|---------------------------|
| test sample               | $70.1*5.64$    | $29.9*5.64$      | 564                       |
| control sample            | $3.5*4.09$     | $96.4*4.09$      | 409                       |
|                           |                |                  |                           |
| total                     | 409.679        |  562.912         | 973         |

the expected ratio is the product of the total of all rows and all columns, devided by the total

expected = $\frac{row tot * col tot}{total}$

In [33]:
Ntot = 973
expected = (564*409*409.679*562.912)/973
## ISN'T THERE A MISTAKE, SHOULD NOT WE DIVIDE OVER THE TOTAL?
sample_values = [[70.1*5.64,29.9*5.64],[3.05*4.09,96.4*4.09]]
chisqstat= lambda N, values, expect : N*((values[0][0]*values[1][1]-values[0][1]*values[1][0])**2)/(expect)

print chisqstat(Ntot, sample_values, expected)


420854.992705


#!!!!!! TODO FOR YOU!
#lets see what the chi sq statistics says about the employment for >6 consecutive months (row 6)

Note: I used a different approach that I could understand more easily - and, of course, is equivalent to the method followed before.
$$Expected = \frac{sum_{col}}{total} . \frac{sum_{row}}{total} . total$$

And then the Chi square statistic is the difference $\Chi = \frac{(Expected - Observed)^2}{Expected}$

Calculate expected values:
+ Expected number of employed people in test sample: $$Expected = \frac{26.6}{973} . \frac{564}{973} . 973 = 15.42$$
+ Expected number of unemployed people in test sample: $$Expected = \frac{173.4}{973} . \frac{564}{973} . 973 = 101.04$$
+ Expected number of employed people in test sample: $$Expected = \frac{26.6}{973} . \frac{564}{973} . 973 = 15.42$$
+ Expected number of employed people in test sample: $$Expected = \frac{26.6}{973} . \frac{564}{973} . 973 = 15.42$$


|employed in subsidized job |     employed      | not employes      |                   
|---------------------------|-------------------|-------------------|---------------------------|
| test sample               |  $14.7 * 5.64$    |   $85.3 * 5.64$   | 564                       |
| control sample            | $11.9 * 4.09$     |  $88.1 * 4.09 $   | 409                       |
|                           |                   |                   |                           |
| total                     |      26.6         |    173.4          | 973                       |

## The next table is the final version of the sample results (in quantities instead of percentages):

|employed in subsidized job |     employed      | not employes      |                   
|---------------------------|-------------------|-------------------|---------------------------|
| test sample               |      $82.91$     |     $481.092$     | $564$                       |
| control sample            |      $48.67$     |     $360.329$     | $409$                       |
|                           |                   |                   |                             |
| total                     |     $131.58$     |     $841.421$     | $973$                       |

Following are the expected results, each element of the table obtained as:
$$Expected = \frac{\sum col}{total}* \sum row$$

|employed in subsidized job |     employed      | not employes      |                   
|---------------------------|-------------------|-------------------|---------------------------|
| test sample               |      $76.28$     |     $487.49$     | $564$                       |
| control sample            |      $55.31$     |     $353.7$     | $409$                       |
|                           |                   |                   |                             |
| total                     |     $131.6$     |     $841.421$     | $973$                       |

In [39]:
expected = [[82.91, 481.092],[48.67,360.329]]
sample_values = [[76.28 , 487.49],[55.31, 353.7]]

chisqstat= sum([(expected[0][0]-sample_values[0][0])**2/expected[0][0],
               (expected[0][1]-sample_values[0][1])**2/expected[0][1],
               (expected[1][0]-sample_values[1][0])**2/expected[1][0],
               (expected[1][1]-sample_values[1][1])**2/expected[1][1]])
print chisqstat

1.64310538586


In [34]:
Image("http://passel.unl.edu/Image/Namuth-CovertDeana956176274/chi-sqaure%20distribution%20table.PNG")

NameError: name 'Image' is not defined

+ #### 432 is hella larger than 3.84?

If both of them are real numbers with its usual metrics, yes ;)

+ #### Why am i mentioning 3.84?

3.84 is the Chi-square statistic for a tolerance of 0.05 and one degree of freedom (N-1). This means that the null hypothesis can be rejected for this level of confidence (or any other in the table).

+ #### How does the 6mo chi square statistic that you derived compare?

1.643 is actually smaller than 3.84 so the confidence would have to be between 10 and 25% for this to work for us, which is usually not the case, so the null hypothesis cannot be rejected.

+ #### Please state what that means in terms of your hypothesis in a markdown cell below!

The null hypothesis cannot be rejected. It makes sense that the hypothesis test rejected the first null hypothesis regarding the , and not this because the percentual difference is very small compared to the samples, which is not the case in the first example.